main.go

```diff
@@ -171,93 +169,57 @@ func initAlpm() (err error) {
 	return
 }
 
-func initAlpmHandle() (err error) {
+func initAlpmHandle() error {
 	if alpmHandle != nil {
-		err = alpmHandle.Release()
-		if err != nil {
+		if err := alpmHandle.Release(); err != nil {
 			return err
 		}
 	}
-
-	alpmHandle, err = alpmConf.CreateHandle()
-	if err != nil {
-		err = fmt.Errorf("Unable to CreateHandle: %s", err)
-		return
+	var err error
+	if alpmHandle, err = alpmConf.CreateHandle(); err != nil {
+		return fmt.Errorf("Unable to CreateHandle: %s", err)
 	}
 
 	alpmHandle.SetQuestionCallback(questionCallback)
-	return
+	return nil
 }
 
-func main() {
-	var status int
-	var err error
-
-	if 0 == os.Geteuid() {
-		fmt.Println("Please avoid running yay as root/sudo.")
-	}
-
-	err = cmdArgs.parseCommandLine()
-	if err != nil {
-		fmt.Println(err)
-		status = 1
-		goto cleanup
-	}
-
-	err = setPaths()
-	if err != nil {
-		fmt.Println(err)
-		status = 1
-		goto cleanup
-	}
-
-	err = initConfig()
-	if err != nil {
-		fmt.Println(err)
-		status = 1
-		goto cleanup
+func cleanup() {
+	if alpmHandle != nil {
+		temp := alpmHandle
+		// set alpmHandle to nil to avoid entering this
+		// branch of code again, at cleanup time.
+		alpmHandle = nil
+		must(temp.Release())
 	}
+}
 
-	cmdArgs.extractYayOptions()
-
-	err = initVCS()
+// must outputs the error if there is one,
+// then calls cleanup() and ends the program with exit code 1.
+// If err == nil, no action is taken.
+func must(err error) {
 	if err != nil {
-		fmt.Println(err)
-		status = 1
-		goto cleanup
-
+		fmt.Fprintln(os.Stderr, err)
+		cleanup()
+		os.Exit(1)
 	}
+}
 
-	err = initAlpm()
-	if err != nil {
-		fmt.Println(err)
-		status = 1
-		goto cleanup
+func main() {
+	if os.Geteuid() == 0 {
+		fmt.Println("Please avoid running yay as root/sudo.")
 	}
 
-	err = handleCmd()
-	if err != nil {
-		if err.Error() != "" {
-			fmt.Println(err)
-		}
-
-		status = 1
-		goto cleanup
-	}
+	// Ensure release of alpmHandle, if os.Exit is not called
+	defer cleanup()
```

Comment:
I'm talking about this cleanup specifically. This could be to the end of the function and it will not change the behaviour. No defer = no possibility of getting confused about os.Exit cancelling it.

Reply:
I agree that the combination of `defer` and `os.Exit` is unfortunate, since `os.Exit` does not trigger the deferred functions. Placing `os.Exit` in the final deferred function would solve this.